cpp/src/copying/gather.cu

```diff
@@ -360,38 +266,104 @@ void gather(table const* source_table, gdf_index_type const gather_map[],
+
+  // We create `n_cols` streams for the `n_cols` kernels we are gonna launch.
+  std::vector<util::cuda::scoped_stream> v_stream(n_cols);
 
-  auto gather_column = [gather_map, check_bounds, stream](
-                           gdf_column const* source, gdf_column* destination) {
-    CUDF_EXPECTS(source->dtype == destination->dtype, "Column type mismatch");
+  for (gdf_size_type i = 0; i < n_cols; i++) {
+    // Perform sanity checks
+    gdf_column *dest_col = destination_table->get_column(i);
+    const gdf_column *src_col = source_table->get_column(i);
 
-    // If the source column has a valid buffer, the destination column must
-    // also have one
-    bool const source_has_nulls{source->valid != nullptr};
-    bool const dest_has_nulls{destination->valid != nullptr};
-    CUDF_EXPECTS((source_has_nulls && dest_has_nulls) || (not source_has_nulls),
-                 "Missing destination validity buffer");
+    CUDF_EXPECTS(src_col->dtype == dest_col->dtype, "Column type mismatch");
 
-    // TODO: Each column could be gathered on a separate stream
-    cudf::type_dispatcher(source->dtype, column_gatherer{}, source, gather_map,
-                          destination, check_bounds, stream);
+    CUDF_EXPECTS(dest_col->data != nullptr, "Missing source data buffer.");
+
+    // The data gather for n columns will be put on the first n streams
+    cudf::type_dispatcher(src_col->dtype, column_gatherer{}, src_col,
+                          gather_map, dest_col, check_bounds, v_stream[i],
+                          merge_nvstring_category);
+
+    if (cudf::is_nullable(*src_col)) {
+      CUDF_EXPECTS(cudf::is_nullable(*dest_col),
+                   "Missing destination null mask.");
+    }
+  }
 
-    return destination;
-  };
+  rmm::device_vector<gdf_size_type> d_count_vec(n_cols, 0);
 
-  // Gather columns one-by-one
-  std::transform(source_table->begin(), source_table->end(),
-                 destination_table->begin(), destination_table->begin(),
-                 gather_column);
+  std::vector<bit_mask_t *> h_bit_src(n_cols);
+  std::vector<bit_mask_t *> h_bit_dest(n_cols);
 
+  std::vector<rmm::device_vector<bit_mask_t>> vec_temp_bit(n_cols);
+
+  // loop over each column, check if inplace and allocate buffer if true.
+  for (gdf_size_type i = 0; i < n_cols; i++) {
+    const gdf_column *dest_col = destination_table->get_column(i);
+    h_bit_src[i] =
+        reinterpret_cast<bit_mask_t *>(source_table->get_column(i)->valid);
+    // Allocate inplace buffer
+    if (cudf::is_nullable(*dest_col) &&
+        dest_col->valid == source_table->get_column(i)->valid) {
+      vec_temp_bit[i].resize(dest_col->size);
+      h_bit_dest[i] = vec_temp_bit[i].data().get();
+    } else {
+      h_bit_dest[i] = reinterpret_cast<bit_mask_t *>(dest_col->valid);
+    }
+  }
+
+  // In the following we allocate the device array thats hold the valid
+  // bits.
+  rmm::device_vector<bit_mask_t *> d_bit_src(n_cols);
```

Comment:
`thrust::device_vector` has a copy constructor from a `std::vector`. No need for the manual memcpy. 
```suggestion
  rmm::device_vector<bit_mask_t *> d_bit_src(h_bit_src);
```